// eb-toolkit/dist/cli/delete-application.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("../index");
const commandModule = {
    describe: 'delete application',
    handler: function ({ appName, region, terminateEnvByForce }) {
        return __awaiter(this, void 0, void 0, function* () {
            const application = new index_1.Application({
                region,
            });
            yield application.deleteApplication(appName, terminateEnvByForce);
        });
    },
    command: 'delete-application',
    builder: {
        appName: {
            demandOption: true,
            type: 'string',
            description: 'Beanstalk application name',
        },
        terminateEnvByForce: {
            demandOption: false,
            type: 'boolean',
            description: 'Whether to terminate Environment by force',
        },
        region: {
            demandOption: false,
            type: 'string',
            description: 'AWS region',
        },
    },
};
exports.default = commandModule;
