// eb-toolkit/dist/cli/describe-environment.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("../index");
const commandModule = {
    describe: 'get environments',
    handler: function ({ appName, appEnv, envName, region }) {
        return __awaiter(this, void 0, void 0, function* () {
            const application = new index_1.Application({
                region,
            });
            if (!envName && appName && appEnv) {
                envName = yield application.getDefaultEnvironment(appName, appEnv);
            }
            if (!envName) {
                throw new Error('No envName was specified');
            }
            const environmentDesc = yield application.environment.describeEnvironment(envName);
            console.info(environmentDesc);
        });
    },
    command: 'describe-env',
    builder: {
        appName: {
            demandOption: false,
            type: 'string',
            description: 'Beanstalk application name',
        },
        appEnv: {
            demandOption: false,
            type: 'string',
            description: 'Beanstalk application env',
        },
        envName: {
            demandOption: false,
            type: 'string',
            description: 'Beanstalk environment name. Either this or (appName and appEnv) is required',
        },
        region: {
            demandOption: false,
            type: 'string',
            description: 'AWS region',
        },
    },
};
exports.default = commandModule;
